### PR TITLE
feat: enhancement for allocate judge (FPLA-1321)

### PR DESCRIPTION
**Pull request by Jamesnelson117**
**Before creating a pull request make sure that:**

- [x] PR's title includes JIRA ticket number and follow [Conventional Commits](https://www.conventionalcommits.org/) specification
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-1321

### Change description ###
Adds option to select allocated judge (if provided) in the following events
- Create an order
- Notice of proceedings
- Draft SDO

Screenshot 2020-04-08 at 11 21 36

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java [hunk 1]
```diff
@@ -87,8 +93,23 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody Call
...
+            if (isNotEmpty(caseData.getStandardDirectionOrder())
+                && isNotEmpty(caseData.getStandardDirectionOrder().getJudgeAndLegalAdvisor())) {
+                judgeAndLegalAdvisor = caseData.getStandardDirectionOrder().getJudgeAndLegalAdvisor();
+            } else {
+                judgeAndLegalAdvisor = JudgeAndLegalAdvisor.builder().build();
+            }
+
+            judgeAndLegalAdvisor.setAllocatedJudgeLabel(buildAllocatedJudgeLabel(caseData.getAllocatedJudge()));
+
+            caseDetails.getData().put(JUDGE_AND_LEGAL_ADIVSOR_KEY, judgeAndLegalAdvisor);
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java:112]: spelling mistake here on advisor

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java [hunk 2]
```diff
@@ -87,8 +93,23 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody Call
...
+
+        if (isNotEmpty(caseData.getAllocatedJudge())) {
+            JudgeAndLegalAdvisor judgeAndLegalAdvisor;
+
+            if (isNotEmpty(caseData.getStandardDirectionOrder())
+                && isNotEmpty(caseData.getStandardDirectionOrder().getJudgeAndLegalAdvisor())) {
+                judgeAndLegalAdvisor = caseData.getStandardDirectionOrder().getJudgeAndLegalAdvisor();
+            } else {
+                judgeAndLegalAdvisor = JudgeAndLegalAdvisor.builder().build();
+            }
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java:108]: Is here the best place for this logic? perhaps we could move this to a private method that returns the judge and legal advisor? what do you think?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java [hunk 3]
```diff
@@ -87,8 +93,23 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody Call
...
             directions.forEach((key, value) -> caseDetails.getData().put(key.getValue(), value));
 
-            caseDetails.getData()
-                .put("judgeAndLegalAdvisor", caseData.getStandardDirectionOrder().getJudgeAndLegalAdvisor());
+            caseDetails.getData().put(JUDGE_AND_LEGAL_ADIVSOR_KEY,
+                caseData.getStandardDirectionOrder().getJudgeAndLegalAdvisor());
+        }
+
+        if (isNotEmpty(caseData.getAllocatedJudge())) {
+            JudgeAndLegalAdvisor judgeAndLegalAdvisor;
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java:101]: could we do JudgeAndLegalAdvisor judgeAndLegalAdvisor = JudgeAndLegalAdvisor.builder().build(); here and remove the need for the else block?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderController.java [hunk 4]
```diff
@@ -84,6 +89,16 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody Call
             caseDetails.getData().put("dateOfIssue", time.now().toLocalDate());
         }
 
+        if (isNotEmpty(caseData.getAllocatedJudge())) {
+            String assignedJudgeLabel = buildAllocatedJudgeLabel(caseData.getAllocatedJudge());
+
+            JudgeAndLegalAdvisor judgeAndLegalAdvisor = JudgeAndLegalAdvisor.builder()
+                .allocatedJudgeLabel(assignedJudgeLabel)
+                .build();
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/GeneratedOrderController.java:97]: Could have a similar setAllocatedJudgeLabel method for this too. :)

File: service/src/test/java/uk/gov/hmcts/reform/fpl/utils/JudgeAndLegalAdvisorHelperTest.java [hunk 5]
```diff
@@ -97,5 +101,85 @@ private void shouldExtractOtherTitleDescriptionWhenOtherTitleSelected() {
...
+
+    @Test
+    void shouldReturnInputtedJudgeWhenUseAllocatedJudgeNotSelected() {
+        JudgeAndLegalAdvisor judgeAndLegalAdvisor = JudgeAndLegalAdvisor.builder()
+            .judgeTitle(MAGISTRATES)
+            .judgeLastName("Stevenson")
+            .legalAdvisorName("John Papa")
+            .useAllocatedJudge("No")
+            .allocatedJudgeLabel("Case assigned to: His Honour Judge Dread")
+            .build();
```
Comment (toireasam) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/utils/JudgeAndLegalAdvisorHelperTest.java:132]: As we use this here & the test above we could extract to private method with parameter useAllocatedJudge set to yes/no depending. 

- if you don't wish to refactor here I don't mind

File: service/src/test/java/uk/gov/hmcts/reform/fpl/utils/JudgeAndLegalAdvisorHelperTest.java [hunk 6]
```diff
@@ -97,5 +101,85 @@ private void shouldExtractOtherTitleDescriptionWhenOtherTitleSelected() {
...
+        assertThat(label).isEqualTo("Case assigned to: His Honour Judge Dread");
+    }
+
+    @Test
+    void shouldBuildAllocatedJudgeLabelWhenAllocatedJudgeIncludesOtherTitleAndFullName() {
+        Judge allocatedJudge = Judge.builder()
+            .judgeTitle(OTHER)
+            .otherTitle("Mr")
+            .judgeLastName("Watson")
+            .build();
```
Comment (toireasam) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/utils/JudgeAndLegalAdvisorHelperTest.java:158]: could be private method buildAllocatedJudge() 

- again not fussed if you don't wish to refactor

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java [hunk 7]
```diff
@@ -73,6 +81,54 @@ void shouldGenerateC6NoticeOfProceedingsDocument() throws Exception {
...
+        CallbackRequest callbackRequest = buildCallbackRequest();
+
+        Map<String, Object> templateData = createTemplatePlaceholders();
+
+        given(noticeOfProceedingsService.getNoticeOfProceedingTemplateData(any()))
+            .willReturn(templateData);
+        given(docmosisDocumentGeneratorService.generateDocmosisDocument(any(), any()))
+            .willReturn(docmosisDocument);
+        given(uploadDocumentService.uploadPDF(any(), any()))
+            .willReturn(document);
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java:101]: This is used in method above too - could be extracted to a before block

File: e2e/tests/judiciaryAdministersCaseAfterSubmission_test.js [hunk 8]
```diff
@@ -70,21 +80,31 @@ Scenario('Judiciary enters hearing details and submits', async (I, caseViewPage,
...
-  I.seeInTab(['Allocated Judge', 'Judge or magistrate\'s title'], 'Her Honour Judge');
-  I.seeInTab(['Allocated Judge', 'Last name'], 'Moley');
+  I.seeEventSubmissionConfirmation(config.administrationActions.createNoticeOfProceedings);
+  await caseViewPage.goToNewActions(config.administrationActions.createNoticeOfProceedings);
+  await createNoticeOfProceedingsEventPage.checkC6A();
+  await createNoticeOfProceedingsEventPage.useAlternateJudge();
+  await createNoticeOfProceedingsEventPage.selectJudgeTitle();
+  await createNoticeOfProceedingsEventPage.enterJudgeLastName('Sarah Simpson');
+  await createNoticeOfProceedingsEventPage.enterLegalAdvisorName('Ian Watson');
+  await I.completeEvent('Save and continue');
```
Comment (toireasam) [inline e2e/tests/judiciaryAdministersCaseAfterSubmission_test.js:105]: do they need the 'await' before these if they aren't async methods?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java (hunk 7 above)
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java:101]: Not quite. In my latest test I do not need to test params so I use any(). While the above test it provides correct arguments.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java [hunk 9]
```diff
@@ -87,15 +93,32 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody Call
...
+        if (isNotEmpty(caseData.getAllocatedJudge())) {
+            caseDetails.getData().put(JUDGE_AND_LEGAL_ADVISOR_KEY, setAllocatedJudgeLabel(caseData));
         }
 
         return AboutToStartOrSubmitCallbackResponse.builder()
             .data(caseDetails.getData())
             .build();
     }
 
+    private JudgeAndLegalAdvisor setAllocatedJudgeLabel(CaseData caseData) {
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersController.java:109]: nice :)

**Pull request merged**